ambry-store/src/main/java/com/github/ambry/store/Journal.java

```diff
@@ -117,6 +117,22 @@ void addEntry(Offset offset, StoreKey key, Long crc) {
     }
   }
 
+  /**
+   * This method should only be called when replica gets sealed. Remove all entries belongs to the auto closed log segment.
+   * It will delay the addEntry method for sealed replica, so the replication performance will be hold while running this method.
+   * @param activeLogSegmentName journal should remove all entries not belongs to active log segment.
+   */
+  void cleanUpJournal(LogSegmentName activeLogSegmentName) {
+    Map.Entry<Offset, StoreKey> earliestEntry = journal.firstEntry();
+    while (!inBootstrapMode && earliestEntry != null
+        && earliestEntry.getKey().getName().compareTo(activeLogSegmentName) < 0) {
+      journal.remove(earliestEntry.getKey());
+      recentCrcs.remove(earliestEntry.getValue());
+      currentNumberOfEntries.decrementAndGet();
+      earliestEntry = journal.firstEntry();
+    }
```

Comment:
If we use `storeWriteLock` to ensure deletes/updates are blocked while we close the last log segment, do we still need this "while" loop?  Can we directly call `journal.clear()` and `recentCrcs.clear()` and  set `currentNumberOfEntries` to 0?

Reply:
You are right, updated the corresponding logic.